### ISSUE TITLE
No rows returned in LEFT JOIN with < or > against table having no rows inserted

### ISSUE THREAD

**Issue by davetapley** [duckdb/duckdb#6341]
### What happens?

See repro below.

Note it only occurs with `<` and `>`, but not `=`, so perhaps an ordering issue?

### To Reproduce

```sql
CREATE TABLE foo (ts TIMESTAMP);
CREATE TABLE bar (ts TIMESTAMP);

INSERT INTO foo VALUES ('2023-01-01 00:00:00');
INSERT INTO foo VALUES ('2023-01-01 00:00:01');

.print no rows in bar, no rows returned for < or >:
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts = bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts < bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts > bar.ts;

.print with row in bar works as expected:
INSERT INTO bar VALUES ('2023-01-01 00:00:01');
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts = bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts < bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts > bar.ts;

.print delete row in bar and still works as expected:
DELETE FROM bar;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts = bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts < bar.ts;
SELECT foo.ts foo, bar.ts bar FROM foo LEFT JOIN bar ON foo.ts > bar.ts;
```

---

```
no rows in bar, no rows returned for < or >:
┌─────────────────────┬───────────┐
│         foo         │    bar    │
│      timestamp      │ timestamp │
├─────────────────────┼───────────┤
│ 2023-01-01 00:00:00 │           │
│ 2023-01-01 00:00:01 │           │
└─────────────────────┴───────────┘
┌───────────┬───────────┐
│    foo    │    bar    │
│ timestamp │ timestamp │
├───────────────────────┤
│        0 rows         │
└───────────────────────┘
┌───────────┬───────────┐
│    foo    │    bar    │
│ timestamp │ timestamp │
├───────────────────────┤
│        0 rows         │
└───────────────────────┘
```

```
with row in bar works as expected:
┌─────────────────────┬─────────────────────┐
│         foo         │         bar         │
│      timestamp      │      timestamp      │
├─────────────────────┼─────────────────────┤
│ 2023-01-01 00:00:01 │ 2023-01-01 00:00:01 │
│ 2023-01-01 00:00:00 │                     │
└─────────────────────┴─────────────────────┘
┌─────────────────────┬─────────────────────┐
│         foo         │         bar         │
│      timestamp      │      timestamp      │
├─────────────────────┼─────────────────────┤
│ 2023-01-01 00:00:00 │ 2023-01-01 00:00:01 │
│ 2023-01-01 00:00:01 │                     │
└─────────────────────┴─────────────────────┘
┌─────────────────────┬───────────┐
│         foo         │    bar    │
│      timestamp      │ timestamp │
├─────────────────────┼───────────┤
│ 2023-01-01 00:00:00 │           │
│ 2023-01-01 00:00:01 │           │
└─────────────────────┴───────────┘
```
```
delete row in bar and still works as expected:
┌─────────────────────┬───────────┐
│         foo         │    bar    │
│      timestamp      │ timestamp │
├─────────────────────┼───────────┤
│ 2023-01-01 00:00:00 │           │
│ 2023-01-01 00:00:01 │           │
└─────────────────────┴───────────┘
┌─────────────────────┬───────────┐
│         foo         │    bar    │
│      timestamp      │ timestamp │
├─────────────────────┼───────────┤
│ 2023-01-01 00:00:00 │           │
│ 2023-01-01 00:00:01 │           │
└─────────────────────┴───────────┘
┌─────────────────────┬───────────┐
│         foo         │    bar    │
│      timestamp      │ timestamp │
├─────────────────────┼───────────┤
│ 2023-01-01 00:00:00 │           │
│ 2023-01-01 00:00:01 │           │
└─────────────────────┴───────────┘
```

### OS:

Linux

### DuckDB Version:

v0.7.1-dev57 8da10c8e33

### DuckDB Client:

CLI

### Full Name:

Dave Tapley

### Affiliation:

JE Fuller

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Mytherin** [comment]
Thanks for the report! I have pushed a fix in #6342. This was caused by an optimizer incorrectly assuming that if all join conditions in a `LEFT` join are true, the join can be replaced with a cross product. That is correct only if neither side is empty - since the cross product returns the empty set if either side is empty, and a left join returns the left set if the right side is empty.

**Comment by davetapley** [comment]
Amazing, thanks @Mytherin.

Between this and #6323 this has got to be the fastest time to fix I've ever seen in an open-source project 😂 🙏🏻
